### PR TITLE
Component Framework

### PR DESCRIPTION
**Pull request by Feiyang1**
Adopt the component framework to manage SDK registration and  inter SDK dependencies.
Design doc: go/firebase-js-platform

### CODE REVIEW

File: packages/messaging/src/controllers/window-controller.ts [hunk 1]
```diff
@@ -308,16 +308,23 @@ export class WindowController extends BaseController {
...
           const eventType = getEventType(firebaseMessagingType);
-          (this.app as _FirebaseApp).INTERNAL.analytics.logEvent(
-            eventType,
-            /* eslint-disable camelcase */
-            {
-              message_name: data[FN_CAMPAIGN_NAME],
-              message_id: data[FN_CAMPAIGN_ID],
-              message_time: data[FN_CAMPAIGN_TIME],
-              message_device_time: Math.floor(Date.now() / 1000)
+          this.services.analyticsProvider.get().then(
```
Comment (mmermerkaya) [inline packages/messaging/src/controllers/window-controller.ts]: Can you make the event listener `async` and use `await` here?

File: packages/messaging/test/testing-utils/make-fake-firebase-services.ts [hunk 2]
```diff
@@ -34,12 +48,23 @@ export function makeFakeApp(options: FirebaseOptions = {}): FirebaseApp {
...
-        delete: () => Promise.resolve()
-      };
-    }
+    installations: null as any
+  };
+}
+
+export function makeFakeInstallations(): FirebaseInstallations {
+  return {
+    getId: () => Promise.resolve('FID'),
```
Comment (mmermerkaya) [inline packages/messaging/test/testing-utils/make-fake-firebase-services.ts]: Very much a nit, but you can do `getId: async () => 'FID',`. Also the other ones.

File: packages/component/README.md [hunk 3]
```diff
@@ -0,0 +1,20 @@
+# @firebase/component
+
+_NOTE: This is specifically tailored for Firebase JS SDK usage, if you are not a
+member of the Firebase team, please avoid using this package_
+
+## Installation
```
Comment (hsubox76) [inline packages/component/README.md]: Should we get rid of the rest of the instructions from here on down if nobody should be installing it anyway?

File: packages/messaging/src/controllers/window-controller.ts (hunk 1 above)
Comment (Feiyang1) [reply packages/messaging/src/controllers/window-controller.ts]: Done.

File: packages/messaging/test/testing-utils/make-fake-firebase-services.ts (hunk 2 above)
Comment (Feiyang1) [reply packages/messaging/test/testing-utils/make-fake-firebase-services.ts]: Nice! Changed here and other places where we return a promise.

File: packages/component/README.md (hunk 3 above)
Comment (Feiyang1) [reply packages/component/README.md]: Good point, but I think it's harmless to keep it, and it's consistent with other packages, e.g. `@firebase/util`.

**Pull request merged**